add Pry::Prompt.select_default

r-obert opened the pull request:
This method can be used to set the default prompt from inside a
.pryrc file, eg:

```ruby
Pry::Prompt.select_default 'nav'
```

or

```ruby
Pry::Prompt.select_default 'rails' # using pry-rails
```

r-obert commented:
@kyrylo 
functionally `Pry::Prompt.select_prompt 'simple'` and `Pry.config.prompt = Pry::Prompt[:simple][:value]` are the same. 

select_prompt is a higher level abstraction that doesn't expose the two-proc array of prompts directly to the user, it requires fewer keystrokes to type, and reduces knowledge of prompts to just the prompts name, so i believe it is easier to use as well.

kyrylo commented:
`Pry::Prompt[:simple][:value]` will still be accessible, though. Consumers must be able to access procs in order to `#call` them. Therefore, this API adds an extra way of doing something that's already possible, which I'd rather avoid.

r-obert commented:
yeah, sure, it would be possible to use both ways. 

the difference is that `select_default` is easier to use, it has a nicer interface, and reduces knowledge of prompts to only knowing a `name`, so it is an improvement over the lower-level API of accessing the two-proc array directly, which `Pry::Prompt` already hides and abstracts away from the user.

As soon as we added `Pry::Prompt.add()` we added another API for prompts, there was already an API for that (`Pry.config.prompt = [proc{..}, proc{..}]`), so this PR is just a continuation of those improvements we already made.

kyrylo commented:
The change makes sense to me but I believe we should do something about `Pry::Prompt[:simple][:value]`. If we have this new method that you added, why would anybody want to call `Pry::Prompt[:simple][:value]`? It's currently public. Will it make sense to deprecate `Pry::Prompt#[]`? I agree that consumers shouldn't care about the internal structure, so I would try and see if can leave only 1 way.

I am thinking whether `Pry.config.prompt = Pry::Prompt[:simple]` would be cleaner here. Note the missing `[:value]` call. `Pry.config.prompt` would know what to call on the passed prompt and we could rename `:value` to whatever we want in the future, or change the implementation completely, while preserving the public API. WDYT?

r-obert commented:
i think `Pry::Prompt[:simple][:value]` spills internal details too much, being able to pass `Pry::Prompt[:simple]`  to `_pry_.prompt=` and `Pry.config.prompt` instead would be nice.

r-obert commented:
it might be better if `Pry::Prompt[:simple]` returned a Struct instead, that encapsulated its name, description, and the two proc arrays.

kyrylo commented:
> it might be better if Pry::Prompt[:simple] returned a Struct instead, that encapsulated its name, description, and the two proc arrays.

That would be no different to what we have now. Users would need to call `Pry::Prompt[:simple].procs` anyway. My idea is to hide that and let Pry figure out what to do with the given prompt object.

r-obert commented:
it would be different, for one it could encapsulate its name.
that information is not encapsulated in the hash we return right now.
anyway i did not propose that as __the__ solution, it was an additional suggestion.

r-obert commented:
it could be useful to keep the name as part of the object we return (struct or not) for cases where the prompt object is passed into other objects/scopes, and the key used for the prompt is not known in those contexts.

kyrylo commented:
I think what we have here is the best solution. I am leaning towards `Pry.config.prompt = Pry::Prompt[:simple]`, still.

> it could be useful to keep the name as part of the object we return 

We don't have a real use case now so I'd rather not do it and let people access prompts through the public API, without the ability to pass it.

r-obert commented:
`Pry.config.prompt = Pry::Prompt[:simple]` is nicer than `Pry.config.prompt = Pry::Prompt[:simple][:value]` for sure, i like `Pry::Prompt.select_default 'simple'` too.

r-obert commented:
i think:
Pry.config.prompt = Pry::Prompt[:simple]
and
Pry::Prompt.select_default 'simple'

are as good as each other for a public API, so i would pick the one with the cleanest implementation.

r-obert commented:
@kyrylo 
updated to use `super`: https://github.com/pry/pry/pull/1860/commits/7b853377cd897924d1ce7bb77d5d4316a3a6b831

r-obert commented:
i think `_pry_.prompt = Pry::Prompt[:simple]` should work as well, wdyt?

r-obert commented on lib/pry/config.rb:
```diff
@@ -31,5 +31,14 @@ def call
     end
+
+    def prompt=(obj)
+      is_prompt_object = lambda do |prompt|
```
@kyrylo 
if prompts were a Struct or similar, we could use `is_a?` here instead.
eg: `Pry::Prompt::PromptInfo === obj`

r-obert replied:
this diff is outdated

r-obert commented:
@kyrylo 
added support for `_pry_.prompt = Pry::Prompt[:nav]` in https://github.com/pry/pry/pull/1860/commits/ff065ba4472b75e92e694b58aa03ba4f7fbfbf5e

r-obert commented:
oh nevermind, i forgot we return the two proc array from that configuration option, kinda confusing.

r-obert commented:
@kyrylo 
here is a summary of this PR currently:

* Removed `Pry::Prompt.select_default`.

* Support `Pry.config.prompt = Pry::Prompt[:nav]`.

* Support `_pry_.prompt = Pry::Prompt[:nav]`.

* By default `Pry.config.prompt` returns a prompt object and not a two-proc array.

* The two proc array has become more of an implementation detail, but the old API still works (eg `_pry_.prompt = [proc{}, proc{}]` and `Pry.config.prompt = [proc{}, proc{}]`.

wdyt?

r-obert commented:
updated my last comment, please re-check.

kyrylo commented:
This PR doesn't pass linting and likely will make some tests fail. I have pushed a PR that doesn't change the internal API https://github.com/pry/pry/pull/1877.

r-obert commented:
linting and specs are minor issues i could have fixed, but okay as you wish.

kyrylo commented:
I didn't mean to replace this PR, just pushed mine to compare.

r-obert commented:
oh i see, sorry. 
lets work from the PR you opened instead.
